Embedded Schemas Intro (#3799)
elixir-ecto/ecto: guides/introduction/Embedded Schemas.md

# Embedded Schemas

Embedded schemas allow you to define and validate structured data. This data can live in memory, or can be stored in the database. Some use cases for embedded schemas include:

- You are maintaining intermediate-state data, like when UI form fields map onto multiple tables in a database.

- You are working within a persisted parent schema and you want to embed data that is...

  - simple, like a map of user preferences inside a User schema.
  - changes often, like a list of product images with associated structured data inside a Product schema.
  - requires complex tracking and validation, like an Address schema inside a User schema.

- You are using a document storage database and you want to interact with and manipulate embedded documents.

## User Profile Example

> Let's explore an example where we have a User and want to store "profile" information about them. The data we want to store is UI-dependent information which is likely to change over time alongside changes in the UI. Also, this data is not necessarily important enough to warrant a new User `field`s in the User schema, as it is not data that is fundamental to the User. An embedded schema is a good solution for this kind of data.

Reviewer: > enough to warrant new User `field`s
or
> enough to warrant a new User `field`

I think the first?

    field :confirmed_at, :naive_datetime

    embeds_one :profile do
      field :online, :boolean
      field :dark_mode, :boolean
      field :visibility, Ecto.Enum, values: [:public, :private, :friends_only]
    end

    timestamps()
  end
end
```

### Embeds

There are two ways to represent embedded data within a schema, `embeds_many`, which creates a list of embeds, and `embeds_one`, which creates only a single instance of the embed. Your choice here affects the behavior of embed-specific functions like `Ecto.Repo.put_embed/4` and `Ecto.Repo.cast_embed/4`, so choose whichever is most appropriate to your use case. In our example we are going to use `embeds_one` since users will only ever have one profile associated with them.

```elixir
defmodule User do
  use Ecto.Schema

  schema "users" do
    field :full_name, :string
    field :email, :string
    field :avatar_url, :string
    field :confirmed_at, :naive_datetime

    embeds_one :profile do
      field :online, :boolean
      field :dark_mode, :boolean
      field :visibility, Ecto.Enum, values: [:public, :private, :friends_only]
    end

    timestamps()
  end
end
```

### Extracting the embeds

> While the above User schema is simple and sufficient, we might want to work independently with the embedded profile struct. For example, if there were a lot of functions devoted solely to manipulating the profile data. In such scenarios, it is recommended to extract the embedded struct into its own schema using the `embedded_schema` function.

Reviewer: > embedded profile struct if, for example, there were...

The separate `For example,...` sentence reads a little funky.

Author: reworked the paragraph based on the feedback <3

# Embedded Schemas

> Embedded schemas allow you to define and validate structured data that is nested within another struct. This data can live exclusively in memory, or can be stored in the database.

Reviewer 2: The name here is confusing but the original definition is correct. You don't need to nest an embedded struct within another struct. In fact, if you have in-memory data, you use an embedded schema too!

Reviewer 2: I find the addition of the word exclusively confusing. I understand the meaning you want to imply but I don't think it will be clear for newcomers. For now I recommend reverting this paragraph and keeping the original introduction. :)

Author: "spiritually reverted" in my next commit with some other options as well

Some use cases for embedded schemas include:

- Embedding flexible data that changes often, like a map of user preferences inside a User schema.

- Embedding simple data that you want to track and validate, but where maintaining a separate table and tracking foreign keys and many-to-many relationships is unnecessary, like a list of product images.

- When you want nuanced control over the tracking and validation of a complex struct, like if you wanted a changeset for address validation, but didn't want a separate address table.

- When using document storage databases, and you want to interact with and manipulate embedded documents.

> These above use cases have a few themes in common. They all have nested data, and it makes sense for that nested data to be (1) bundled together into a sub-entity, and (2) changeset validations on that sub-entity are desireable.

Reviewer 2: As mentioned above, it is not always about nesting, so we should revisit this. :)

Author: I'm happy to abandon the "nesting" phraseology. Thank you for the feedback <3

I think the thing I want to succinctly articulate is the following: The schema is representing something. That something is not well represented by a flat list of data fields; it is better represented by structured data wherein data lives at multiple layers. From this fact, it logically follows that you will want to handle those sub-structs using their own schemas and changesets. And here comes embedded schemas to save the day!

Do you agree with that core narrative? If so, then I will continue trying to find a simple and succinct way to articulate it. Otherwise, I need to keep reading the docs XD

Reviewer 2: That's a good chunk of the narrative but the truth is that embedded schemas can completely exist on their own, even if not nested anywhere, as they are useful to validate data that is not persisted anywhere. This guide has a brief example of it: https://github.com/elixir-ecto/ecto/blob/master/guides/howtos/Data%20mapping%20and%20validation.md

Author: That is _really_ helpful feedback. Thank you!! You give a similar example here, which was a nice explanation as well: https://hexdocs.pm/ecto/Ecto.Schema.html#module-example . 

I reckon others get confused by the name `embedded_schema` the same way I did, and focus solely on the "nesting" aspects of it at first. I think I'd like to call that out somewhere in this intro to help future readers, if you think that's a good idea.

In trying to understand why you couldn't use a regular schema in places where you use embedded_schema, I realized that the programmatic difference between them is really that schemas have the __meta__ map that keeps track of persistence state. So it's really that `embedded_schema` == `persistence_agnostic_schema`, right? You get all the upside of schemas without worry about persistence. And if you _do_ want to worry about persistence, you do so via a parent schema, or manually like you did in the Data mapping and validation guide.

Reviewer 2: You got it!

Author: <3 I'm adding the section at the ended of the "extracting" section to mention this, so resolving it here.

## Example

Let's explore an example where we have a User and want to store "profile" information about them. The data we want to store here is a loose grouping of UI-dependent information, which is likely to change over time alongside changes in the UI. Also, this data is not necessarily important enough to warrant a new User `field` in the schema, as it is not data that is fundamental to the User. An embedded schema is a good solution for this kind of data.

```elixir
defmodule User do
  use Ecto.Schema

  schema "users" do
    field :full_name, :string
    field :email, :string
    field :avatar_url, :string
    field :confirmed_at, :naive_datetime

    embeds_one :profile do
      field :online, :boolean
      field :dark_mode, :boolean

>       field :visibility, Ecto.Enum, values: [:public, :private, :friends_only]

Reviewer 2: Love these updated fields!

    end

    timestamps()
  end
end
```

### `embeds_one` and `embed_many`

One of the first choices to make is how to represent the embedded data within the struct. Do we want to store an array of structs using `embeds_many`, or just one using `embeds_one`? In our example we are going to use `embeds_one` since users will only ever have one profile associated with them.

```elixir
defmodule User do
  use Ecto.Schema

  schema "users" do
    field :full_name, :string
    field :email, :string
    field :avatar_url, :string
    field :confirmed_at, :naive_datetime

    embeds_one :profile do
      field :online, :boolean
      field :dark_mode, :boolean
      field :visibility, Ecto.Enum, values: [:public, :private, :friends_only]
    end

    timestamps()
  end
end
```

### Extracting embeds

> While the above User schema is simple and sufficient, you might find yourself in a situation where you want to work independently with the embedded profile struct. In such scenarios, it is recommended to extract the embedded struct into it's own schema using the `embedded_schema` function.

Reviewer 2: My answer is: when you start adding a bunch of functions to work exclusively with that embedded schema, then you should extract it.

It is also required for the non-nested cases. :)

> While the above User schema is simple and sufficient, you might find yourself in a situation where you want to work independently with the embedded profile struct. In such scenarios, it is recommended to extract the embedded struct into it's own schema using the `embedded_schema` function.

Author: I mention what you suggest in my next commit.

# Embedded Schemas

**opening paragraph option 2**
Embedded schemas are schemas without persistence. In fact, you can think of `embedded_schema` as `persistence_agnostic_schema`. (Using an embedded schema does not _prevent_ you from persisting the data, it just doesn't concern itself with that task.) This makes embedded schemas ideal for scenarios where you want to manage structured data without necessarily persisting it. For example, when you want an intermediate data structure between your UI layer and your data layer, or when you want to independently manipulate embedded structs as sub-schemas inside a parent schema. Some use cases for embedded schemas include:

**opening paragraph option 1; a 'spiritual reversion' to the original**

> Embedded schemas allow you to define and validate structured data. This data can live in memory, or can be stored in the database. Some use cases for embedded schemas include:

Author: I have created two options. Option 1 is inline with the original opening. Option 2 will probably just be deleted soon, it's a wilder opening based on my interpretation of the readings you gave me. I want to at least get your thoughts on it, even though it's probably a bad opening. Intros are notoriously hard, so will probably just save the final edits for last.

Reviewer 2: Option 2 brings good points but that's not how I would start the discussion with. Rather, when you move the embedded schema to its own module, I would have a paragraph explaining that `embedded_schemas` in fact do not have to live inside another schema at all, and they are great for when you don't need to persist the data anywhere too.

Author: going with option 1 and bringing the option 2 info to the "extraction section

  - requires complex tracking and validation, like an Address schema inside a User schema.

- You are using a document storage database and you want to interact with and manipulate embedded documents.

## User Profile Example

Let's explore an example where we have a User and want to store "profile" information about them. The data we want to store is UI-dependent information which is likely to change over time alongside changes in the UI. Also, this data is not necessarily important enough to warrant a new User `field`s in the User schema, as it is not data that is fundamental to the User. An embedded schema is a good solution for this kind of data.

```elixir
defmodule User do
  use Ecto.Schema

  schema "users" do
    field :full_name, :string
    field :email, :string
    field :avatar_url, :string
    field :confirmed_at, :naive_datetime

    embeds_one :profile do
      field :online, :boolean
      field :dark_mode, :boolean
      field :visibility, Ecto.Enum, values: [:public, :private, :friends_only]
    end

    timestamps()
  end
end
```

### `embeds_one` and `embeds_many`

> There are two ways to represent embedded data within a schema, `embeds_many`, which creates a List of embeds, and `embeds_one`, which creates only a single instance of the embed. Your choice here affects the behavior of embed-specific functions like `Ecto.Repo.put_embed/4` and `Ecto.Repo.cast_embed/4`, so choose whichever is most appropriate to your use case. In our example we are going to use `embeds_one` since users will only ever have one profile associated with them.

Author: I elaborated a little more here on `embeds_many` and `embeds_one`

defmodule UserProfile do
  use Ecto.Schema

  embedded_schema do
    field :online, :boolean
    field :dark_mode, :boolean
    field :visibility, Ecto.Enum, values: [:public, :private, :friends_only]
  end
end
```

> It is important to remember that `embedded_schema` has many use cases independent of `embeds_one` and `embeds_many`. **Embedded schemas are schemas without persistence.** You can think of them as a persistence agnostic `schema`. This makes embedded schemas ideal for scenarios where you want to manage structured data without necessarily persisting it. For example, when you want an intermediate data structure, or when you want to independently work with structured data inside a parent schema.

Reviewer 2: ```suggestion
It is important to remember that `embedded_schema` has many use cases independent of `embeds_one` and `embeds_many`. You can think of them as a persistence agnostic `schema`. This makes embedded schemas ideal for scenarios where you want to manage structured data without necessarily persisting it. For example, if you want to build a contact form, you still want to parse and validate the data, but the data is likely not persisted anywhere. Instead, it is used to send an email. Embedded schemas would be a good fit for such use case.
```

> It is important to remember that `embedded_schema` has many use cases independent of `embeds_one` and `embeds_many`. **Embedded schemas are schemas without persistence.** You can think of `embedded_schema` as something like a persistence agnostic schema. This makes embedded schemas ideal for scenarios where you want to manage structured data without necessarily persisting it. For example, when you want an intermediate data structure between your UI layer and your data layer, or when you want to independently manipulate embedded structs inside a parent schema.

Reviewer: > For example... inside a parent schema, an `embedded_schema` is a good choice or something something.

Author: I feel like this is an awkward sounding paragraph. I tried to rework it inspired by your suggestion. Not exactly how you suggested. It's a really important point I think, for new ecto devs to learn, so I want it to be a good paragraph, but idk, writing's hard, lol. If you have more edits here from what I changed it to, would love to heard them :D

### Migrations

If you wish to save your embedded schema to the database, you need to write a migration to include the embedded data.

```elixir
alter table("users") do
  add :profile, :map
end
```

> Whether you use `embeds_one` or `embeds_many`, it is recommended to use the `:map` data type (although `{:array, :map}` will work with `embeds_many` as well). The reason is that typical relational databases (like Postgres) are likely to represent a `:map` as JSON or JSONB, allowing Ecto adapter libraries more flexibility over how to efficiently store the data.

Reviewer 2: ```suggestion
Whether you use `embeds_one` or `embeds_many`, it is recommended to use the `:map` data type (although `{:array, :map}` will work with `embeds_many` as well). The reason is that typical relational databases are likely to represent a `:map` as JSON (or JSONB in Postgres), allowing Ecto adapter libraries more flexibility over how to efficiently store the data.
```

> Whether you use `embeds_one` or `embeds_many`, it is recommended to use the `:map` data type (although `{:array, :map}` will work with `embeds_many` as well). The reason is that typical relational databases (like Postgres) are likely to represent a `:map` as JSON or JSONB, allowing Ecto adapter libraries more flexibility over how to represent the data efficiently.

Reviewer: > allowing Ecto adapter libraries more flexibility over how to efficiently store the data.

### Changesets

Changeset functionality for embeds will allow you to enforce arbitrary validations on the data. You can define a changeset function for each module. For example, the UserProfile module could require the `online` and `visibility` fields to be present when generating a changeset.

```elixir
defmodule UserProfile do
  # ...

  def changeset(%UserProfile{} = profile, attrs \\ %{}) do
    profile
    |> cast(attrs, [:online, :dark_mode, :visibility])
    |> validate_required([:online, :visibility])
  end
end

profile = %UserProfile{}
UserProfile.changeset(profile, %{online: true, visibility: :public})
```

> Meanwhile, the User changeset function can require it's own validations without worrying about the details of the UserProfile changes because it can pass that responsibility to UserProfile via `cast_embed/3`. A validation failure in an embed will cause the parent changeset to be invalid, even if the parent changeset itself had no errors.

Reviewer: > its own

(no apostrophe)

changeset.valid? # => true
```

In situations where you have kept the embedded schema within the parent module, e.g., you have not extracted a UserProfile, you can still have custom changeset functions for the embedded data within the parent schema.

```elixir
defmodule User do
  use Ecto.Schema

  schema "users" do
    field :full_name, :string
    field :email, :string
    field :avatar_url, :string
    field :confirmed_at, :naive_datetime

    embeds_one :profile, Profile do
      field :online, :boolean
      field :dark_mode, :boolean
      field :visibility, Ecto.Enum, values: [:public, :private, :friends_only]
    end

    timestamps()
  end

  def changeset(%User{} = user, attrs \\ %{}) do
    user
    |> cast(attrs, [:full_name, :email])
    |> cast_embed(:profile, required: true, with: &profile_changeset/2)
  end

  def profile_changeset(profile, attrs \\ %{}) do
    profile
    |> cast(attrs, [:online, :dark_mode, :visibility])
    |> validate_required([:online, :visibility])
  end
end

changeset = User.changeset(%User{}, %{profile: %{online: true, visibility: :public}})
changeset.valid? # => true
```

> ### Querying embedded data

Author: I went ahead and wrote this query section even though I haven't gotten the `where` queries to work yet (which I mention above in the unresolved comment thread we are chatting in). The `select` queries work just fine. I think once that final issue above gets resolved it's ready to go! (I might do more edit passes, though, because I never like what I write, lol, XD). Thank you for taking the time to help me as I wrote this. It's been an honor <3. I'd love to write more if there's other guides and stuff you've been wanting to write but haven't had time for, even outside ecto, etc. :D

Once you have written embedded data to the database, you can use it in queries on the parent schema.

```elixir
user_changeset = User.changeset(%User{}, %{profile: %{online: true, visibility: :public}})
{:ok, _user} = Repo.insert(user_changeset)

(Ecto.Query.from u in User, select: {u.profile["online"], u.profile["visibility"]}) |> Repo.one
# => {true, "public"}

(Ecto.Query.from u in User, select: u.profile, where: u.profile["visibility"] == :public) |> Repo.all
# => [
#  %UserProfile{
#    id: "...",
#    online: true,
#    dark_mode: nil,
#    visibility: :public
#  }
#]
```

> In typical databases where `:map`s are mapped to JSONB (like Postgres), Ecto is constructing the appropriate jsonpath queries for you. To see more examples of valid syntax for embedded schemas, see [`json_extract_path/2`](https://hexdocs.pm/ecto/Ecto.Query.API.html#json_extract_path/2).

Reviewer: > In databases where `:map`s are stored as JSONB (like Postgres)

Author: Reworked the paragraph and incorporated the feedback, <3

defmodule UserProfile do
  use Ecto.Schema

  embedded_schema do
      field :online, :boolean
      field :dark_mode, :boolean
      field :visibility, Ecto.Enum, values: [:public, :private, :friends_only]
  end
end
```

### Migrations

In order to save embedded schemas to the database you need to write a migration for the embedded data.

```elixir
alter table("users") do
  add :profile, :map
end
```

> Whether you use `embeds_one` or `embeds_many`, it is recommended to use the `:map` data type (although `{:array, :map}` will work with `embeds_many` as well). The reason is that typical relational databases (like Postgres) are likely to represent a `:map` as JSON or JSONB, allowing Ecto adapter libraries more flexibility over how to represent the data, while using `{:array, :map}` requires Ecto adapter libraries to conform more strictly to the databases representation of arrays which could lead to unpredicatable, database-dependent behaviors.

Reviewer 2: Looks great, I think we can make it shorter too:

```suggestion
Whether you use `embeds_one` or `embeds_many`, it is recommended to use the `:map` data type (although `{:array, :map}` will work with `embeds_many` as well). The reason is that typical relational databases (like Postgres) are likely to represent a `:map` as JSON or JSONB, allowing Ecto adapter libraries more flexibility over how to represent the data efficiently.
```

```elixir
defmodule User do
  use Ecto.Schema

  schema "users" do
    field :full_name, :string
    field :email, :string
    field :avatar_url, :string
    field :confirmed_at, :naive_datetime

    embeds_one :profile, Profile do
      field :online, :boolean
      field :dark_mode, :boolean
      field :visibility, Ecto.Enum, values: [:public, :private, :friends_only]
    end

    timestamps()
  end

  def changeset(%User{} = user, attrs \\ %{}) do
    user
    |> cast(attrs, [:full_name, :email])
    |> cast_embed(:profile, required: true, with: &profile_changeset/2)
  end

  def profile_changeset(profile, attrs \\ %{}) do
    profile
    |> cast(attrs, [:online, :dark_mode, :visibility])
    |> validate_required([:online, :visibility])
  end
end

changeset = User.changeset(%User{}, %{profile: %{online: true, visibility: :public}})
changeset.valid? # => true
```

### Querying embedded data

Once you have written embedded data to the database, you can use it in queries on the parent struct.

<!-- TODO: Actually proves this works with local test data. Otherwise, user JSON query fragments -->
```elixir

> from u in User, where: u.profile.dark_mode == true

Author: Do we want to add a conclusion of some sort, and if so, what would be in it?

> from u in User, where: u.profile.dark_mode == true

Reviewer 2: I am not sure what to say. The above should just work on more recent Ecto versions. What version you are using?

> from u in User, where: u.profile.dark_mode == true

Author: @josevalim sorry I tried to get it working but can't. Here's the error:

```
** (Ecto.Query.CompileError) cannot fetch field `online` from `p.profile`. Can only fetch fields from:

  * sources, such as `p` in `from p in Post`
  * named bindings, such as `as(:post)` in `from Post, as: :post`
  * parent named bindings, such as `parent_as(:post)` in a subquery

    (ecto 3.7.1) expanding macro: Ecto.Query.where/3
    iex:19: (file)
    (ecto 3.7.1) expanding macro: Ecto.Query.from/2
    iex:19: (file)
```

Here's some links that led me to believe this is not possible:

* https://groups.google.com/g/elixir-ecto/c/Z20aGxm3IhE
* https://stackoverflow.com/questions/63930941/queriyng-embedded-schema-without-using-fragment

> from u in User, where: u.profile.dark_mode == true

Reviewer 2: Correct. I got the syntax messed up. You should do this: `p.profile["dark_mode"]`.

> from u in User, where: u.profile.dark_mode == true

Author: Ecto.Query.from is not converting to the correct jsonpath syntax, it seems. In the code below first query is healthy, second one fails because it's converting to jsonpath using `#>` instead of `##>` which would allow for the text comparison to succeed.

```
iex(22)> (Ecto.Query.from p in Person, where: p.age == 12) |> Repo.all

07:58:48.330 [debug] QUERY OK source="people" db=1.8ms idle=1095.6ms
SELECT p0."id", p0."first_name", p0."last_name", p0."age", p0."profile" FROM "people" AS p0 WHERE (p0."age" = 12) []
[
  %Friends.Person{
    __meta__: #Ecto.Schema.Metadata<:loaded, "people">,
    age: 12,
    first_name: nil,
    id: 1,
    last_name: nil,
    profile: %Friends.Person.Profile{
      dark_mode: nil,
      id: "b7c1fdef-bdfe-4303-8dfa-830ae60c1fe5",
      online: "yeah"
    }
  }
]
iex(23)> (Ecto.Query.from p in Person, where: p.profile["online"] == "yeah") |> Repo.all

07:59:06.929 [debug] QUERY ERROR source="people" db=0.0ms queue=1.3ms idle=1695.6ms
SELECT p0."id", p0."first_name", p0."last_name", p0."age", p0."profile" FROM "people" AS p0 WHERE ((p0."profile"#>'{"online"}') = 'yeah') []
** (Postgrex.Error) ERROR 22P02 (invalid_text_representation) invalid input syntax for type json

    query: SELECT p0."id", p0."first_name", p0."last_name", p0."age", p0."profile" FROM "people" AS p0 WHERE ((p0."profile"#>'{"online"}') = 'yeah')

Token "yeah" is invalid.
    (ecto_sql 3.7.1) lib/ecto/adapters/sql.ex:760: Ecto.Adapters.SQL.raise_sql_call_error/1
    (ecto_sql 3.7.1) lib/ecto/adapters/sql.ex:693: Ecto.Adapters.SQL.execute/5
    (ecto 3.7.1) lib/ecto/repo/queryable.ex:219: Ecto.Repo.Queryable.execute/4
    (ecto 3.7.1) lib/ecto/repo/queryable.ex:19: Ecto.Repo.Queryable.all/3
```

I can generates this same error in psql:

```
tecto_repo=# select profile from people where profile #>> '{"online"}' = 'yeah';
                                       profile                                       
-------------------------------------------------------------------------------------
 {"id": "b7c1fdef-bdfe-4303-8dfa-830ae60c1fe5", "online": "yeah", "dark_mode": null}
(1 row)

tecto_repo=# select profile from people where profile #> '{"online"}' = 'yeah';
ERROR:  invalid input syntax for type json
LINE 1: ... profile from people where profile #> '{"online"}' = 'yeah';
                                                                ^
DETAIL:  Token "yeah" is invalid.
CONTEXT:  JSON data, line 1: yeah
```

Maybe it's a Postgrex issue, or maybe I'm missing something. I can't find if there's documentation on this feature. (Maybe I found the next guide to write, XD).

> from u in User, where: u.profile.dark_mode == true

Reviewer 2: How is your migration? Did you define the field as :json or :jsonb?

> from u in User, where: u.profile.dark_mode == true

Author: ```
tecto_repo=# \d people
                                      Table "public.people"
   Column   |          Type          | Collation | Nullable |              Default               
------------+------------------------+-----------+----------+------------------------------------
 id         | bigint                 |           | not null | nextval('people_id_seq'::regclass)
 first_name | character varying(255) |           |          | 
 last_name  | character varying(255) |           |          | 
 age        | integer                |           |          | 
 profile    | jsonb                  |           |          | 
Indexes:
    "people_pkey" PRIMARY KEY, btree (id)
```

> from u in User, where: u.profile.dark_mode == true

Author: ```elixir
    alter table :people do
      add :profile, :map
    end
```

defmodule UserProfile do
  use Ecto.Schema

  embedded_schema do
    field :online, :boolean
    field :dark_mode, :boolean
    field :visibility, Ecto.Enum, values: [:public, :private, :friends_only]
  end
end
```

> It is important to remember that `embedded_schema` has many use cases independent of `embeds_one` and `embeds_many`. There is a clear relationship between them, but keep in mind that `embedded_schema` has broader applications, like in the intermediate-state use case from the introduction.

Author: Added a NB so readers recall that embedded_schema is not _only_ useful with embedded data. Something that I surmise trips up new Ecto users.

Reviewer 2: Yes, so from what I said above, I would expand on those use cases here, and not in the intro. :)
